manager/volumes/volumes.go

```diff
@@ -0,0 +1,78 @@
+package volumes
+
+import (
+	"sync"
+
+	"github.com/docker/go-events"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/manager/state/store"
+	// "github.com/container-storage-interface/spec/lib/go/csi"
+)
+
+type VolumeManager struct {
+	store *store.MemoryStore
+
+	// synchronization for starting and stopping the VolumeManager
+	startOnce sync.Once
+
+	stopChan chan struct{}
+	stopOnce sync.Once
+	doneChan chan struct{}
```

Comment:
The two `sync.Once`' expose some oddities in the ordering of `.Run()` and `.Stop()`.  Realistically, running `.Stop()` before `.Run()` wouldn't normally be done, but in this case it appears that `.Stop()` would block.

Is there a reason why `context.WithCancel` wasn't used -- you will get the multiple `.Stop()` + `sync.Once` safety for free.

Reply:
Oh jeez. Ok, so this code follows a very basic pattern that's seen, more or less, through a lot of the system. It's going to get overhauled a lot as we work on it.